pkg/microservices/microservices.go

```diff
@@ -0,0 +1,124 @@
+package microservices
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/golang/glog"
+	"github.com/hobbyfarm/gargantua/pkg/util"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+type MicroService struct {
+	name string
+}
+
+var (
+	AuthN      = MicroService{"authn-service"}
+	AuthR      = MicroService{"authr-service"}
+	User       = MicroService{"user-service"}
+	Rbac       = MicroService{"rbac-service"}
+	AccessCode = MicroService{"accesscode-service"}
+	Setting    = MicroService{"setting-service"}
+	Unknown    = MicroService{""}
+)
+
+func (m MicroService) getUrl() (string, error) {
+	// Create a Kubernetes clientset
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return "", fmt.Errorf("error retrieving InClusterConfig")
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return "", fmt.Errorf("error creating kubernetes clientset")
+	}
+
+	ctx := context.Background()
+
+	// Get the endpoints object for the service
+	endpoints, err := clientset.CoreV1().Endpoints(util.GetReleaseNamespace()).Get(ctx, m.name, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("error retrieving endpoints")
+	}
+
+	var grpcPort int32 = 0
+
+	// Get the grpc port number
+	for _, subset := range endpoints.Subsets {
+		for _, port := range subset.Ports {
+			if port.Name == "grpc" {
+				grpcPort = port.Port
+			}
+		}
+	}
+
+	if grpcPort == 0 {
+		return "", fmt.Errorf("no grpc port found for service %s", m.name)
+	}
+
+	// if port < 1 || port > 65535 {
+	// 	return "", errors.New("Invalid port(" + strconv.Itoa(port) + ")! Port must be within the range 1 to 65535")
+	// }
+	return "dns://" + util.GetK8sDnsServer() + "/" + m.name + "." + util.GetReleaseNamespace() + ".svc.cluster.local:" + strconv.Itoa(int(grpcPort)), nil
```

Comment:
Is it possible to avoid doing a lookup of the kubernetes dns server here? I feel like we should just be able to attempt dialing `rbac-service.[namespace].svc.cluster.local` without trying resolution first.